src/FwdState.cc

```diff
@@ -121,8 +121,10 @@ void
 FwdState::closeServerConnection(const char *reason)
 {
     debugs(17, 3, "because " << reason << "; " << serverConn);
-    comm_remove_close_handler(serverConn->fd, closeHandler);
-    closeHandler = NULL;
+    if (closeHandler) {
```

Comment:
Feels wrong that we may now have serverConn without closeHandler. Can we keep serverConn nil until we can set the closeHandler?

Reply:
Requires to send the serverConnection object in Http::Tunneler::answer and in Security::EncryptorAnswer.

I must note we should include also the closed connection to the answers to give the chance to FwdState to call fwdPconnPool->noteUses(fd_table[serverConnection->fd].pconn.uses) on closed connection.

Does this sounds OK?

Reply:
> Alex: Can we keep serverConn nil until we can set the closeHandler?
> Christos: Requires ...

Not setting serverConn until it is ready to use (and until we can schedule closeHandler) sounds good to me.

> we should include also the closed connection to the answers to give the chance to FwdState to call fwdPconnPool->noteUses(fd_table[serverConnection->fd].pconn.uses) on closed connection.

If we have to send the closed Connection to the initiator for some valid reason, let's send it (and document that reason), but I am not sure such sending would help because Connection::fd of a closed Connection will be negative. Please check whether Tunneler itself can call noteUses() and such _before_ closing the connection it owns. I think it is much better to pass fwdPconnPool (and other pools) to Tunneler than to store/send pconn.uses of a closed connection to the initiator!

But check that fwdPconnPool and such cannot become stale/disappear. We may need to add refconting if they can.

Reply:
commits b1b730db and 756945e9.
The second commit removes the fwdPconnPool->noteUses calls for Http::Tunneler and PeerConnector cases. Currently they are not using pcons so it feels wrong to update fwdPconnPool statistics.

Reply:
> Currently [Http::Tunneler and PeerConnector] are not using pcons so it feels wrong to update fwdPconnPool statistics.

They should update fwdPconnPool statistics if the connection they are working on comes from fwdPconnPool. Does it come from that (or any other) connection pool?